driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -177,53 +168,54 @@ private BsonDocument asGetMoreCommandDocument() {
         return document;
     }
 
+    private void killCursorOnClose() {
+        final ServerCursor localCursor = getCursorForKillCursorOnClose();
+        if (localCursor != null) {
```

Comment:
if localCursor == null the connectionSource is never released.  There seem to be tests for this code path but no assertions about the connectionSource reference count.